### PR TITLE
Continue without setting TCCL

### LINKED ISSUE #303: SecurityException in XMLParsingServiceFactory.getService() / lsp4e
```
org.osgi.framework.ServiceException: Exception in org.eclipse.osgi.internal.framework.XMLParsingServiceFactory.getService()
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.factoryGetService(ServiceFactoryUse.java:228)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.getService(ServiceFactoryUse.java:114)
	at org.eclipse.osgi.internal.serviceregistry.ServiceConsumer$2.getService(ServiceConsumer.java:48)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.getService(ServiceRegistrationImpl.java:568)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.getService(ServiceRegistry.java:542)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.getService(BundleContextImpl.java:660)
	at org.eclipse.core.runtime.ServiceCaller.lambda$7(ServiceCaller.java:388)
	at java.base/java.util.Optional.map(Optional.java:260)
	at org.eclipse.core.runtime.ServiceCaller.getCurrent(ServiceCaller.java:387)
	at org.eclipse.core.runtime.ServiceCaller.callOnce(ServiceCaller.java:158)
	at org.eclipse.core.runtime.ServiceCaller.callOnce(ServiceCaller.java:140)
	at org.eclipse.core.internal.content.XMLRootHandler.parseContents(XMLRootHandler.java:158)
	at org.eclipse.core.runtime.content.XMLRootElementContentDescriber2.fillContentProperties(XMLRootElementContentDescriber2.java:206)
	at org.eclipse.core.runtime.content.XMLRootElementContentDescriber2.checkCriteria(XMLRootElementContentDescriber2.java:135)
	at org.eclipse.core.runtime.content.XMLRootElementContentDescriber2.describe(XMLRootElementContentDescriber2.java:174)
	at org.eclipse.core.internal.content.ContentTypeCatalog.describe(ContentTypeCatalog.java:271)
	at org.eclipse.core.internal.content.ContentTypeCatalog.collectMatchingByContents(ContentTypeCatalog.java:246)
	at org.eclipse.core.internal.content.ContentTypeCatalog.internalFindContentTypesFor(ContentTypeCatalog.java:470)
	at org.eclipse.core.internal.content.ContentTypeCatalog.internalFindContentTypesFor(ContentTypeCatalog.java:529)
	at org.eclipse.core.internal.content.ContentTypeCatalog.findContentTypesFor(ContentTypeCatalog.java:363)
	at org.eclipse.core.internal.content.ContentTypeMatcher.findContentTypesFor(ContentTypeMatcher.java:57)
	at org.eclipse.lsp4e.LSPEclipseUtils.getDocumentContentTypes(LSPEclipseUtils.java:1284)
	at org.eclipse.lsp4e.LanguageServiceAccessor.getLSWrappers(LanguageServiceAccessor.java:291)
	at org.eclipse.lsp4e.LanguageServers$LanguageServerDocumentExecutor.getServers(LanguageServers.java:272)
	at org.eclipse.lsp4e.LanguageServers.executeOnServers(LanguageServers.java:403)
	at org.eclipse.lsp4e.LanguageServers.collectAll(LanguageServers.java:76)
	at org.eclipse.lsp4e.LanguageServers.collectAll(LanguageServers.java:58)
	at org.eclipse.lsp4e.ConnectDocumentToLanguageServerSetupParticipant.lambda$0(ConnectDocumentToLanguageServerSetupParticipant.java:60)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: java.lang.SecurityException: setContextClassLoader
	at java.base/java.util.concurrent.ForkJoinWorkerThread$InnocuousForkJoinWorkerThread.setContextClassLoader(ForkJoinWorkerThread.java:210)
	at org.eclipse.osgi.internal.framework.XMLParsingServiceFactory.getService(XMLParsingServiceFactory.java:53)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse$1.run(ServiceFactoryUse.java:218)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.factoryGetService(ServiceFactoryUse.java:215)
	... 34 more

```
```

eclipse.buildId=4.29.0.I20230814-1800
java.version=17.0.8
java.vendor=Eclipse Adoptium
BootLoader constants: OS=win32, ARCH=x86_64, WS=win32, NL=de_DE
Command-line arguments:  -os win32 -ws win32 -arch x86_64

```

What i did: search for "1" in *.xml and press "Show next match" multiple times - was not reproducible that way though
image

I have seen similar exception in a previous report:
https://bugs.eclipse.org/bugs/show_bug.cgi?id=569286#c24

### PR DESCRIPTION
**Pull request by guw**
When called from within a fork-join pool, setting the TCCL is not permitted. Therefore catch and continue without setting a TCCL.

Fixes #303.

### CODE REVIEW

**Comment by guw** [comment]
@tjwatson Any guidance/requirements for unit testing this?

File: bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/BundleContextImpl.java [hunk 1]
```diff
@@ -846,8 +846,18 @@ Object setContextFinder() {
 		ClassLoader previousTCCL = currentThread.getContextClassLoader();
 		ClassLoader contextFinder = container.getContextFinder();
 		if (previousTCCL != contextFinder) {
-			currentThread.setContextClassLoader(container.getContextFinder());
-			return previousTCCL;
+			try {
+				currentThread.setContextClassLoader(container.getContextFinder());
+				return previousTCCL;
+			} catch (SecurityException e) {
```
Comment (tjwatson) [inline bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/BundleContextImpl.java:852]: I think we should catch `RuntimeException` here.  Given that the security manager deprecation/removal I look forward to the Java team wanting to throw some kind of `RuntimeException` here in the future instead.  I think that would be more safe, besides that, any reason for failing to set the TCCL here should not prevent the rest of the code flow.

File: bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/XMLParsingServiceFactory.java [hunk 2]
```diff
@@ -46,14 +46,22 @@ public Object getService(Bundle bundle, ServiceRegistration<Object> registration
...
+		boolean restoreTccl = true;
 		try {
 			BundleWiring wiring = bundle.adapt(BundleWiring.class);
 			ClassLoader cl = wiring == null ? null : wiring.getClassLoader();
-			if (cl != null)
-				Thread.currentThread().setContextClassLoader(cl);
+			if (cl != null) {
+				try {
+					Thread.currentThread().setContextClassLoader(cl);
+				} catch (SecurityException e) {
```
Comment (tjwatson) [inline bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/XMLParsingServiceFactory.java:56]: Change to `RuntimeException`

**Review by tjwatson — CHANGES_REQUESTED** [review]

File: bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/BundleContextImpl.java (hunk 1 above)
Comment (guw) [inline bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/BundleContextImpl.java:852]:
```suggestion
			} catch (RuntimeException e) {
```

File: bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/XMLParsingServiceFactory.java (hunk 2 above)
Comment (guw) [inline bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/framework/XMLParsingServiceFactory.java:56]:
```suggestion
				} catch (RuntimeException e) {
```

**Review by tjwatson — APPROVED** [review]

**Pull request merged**